### PR TITLE
Add loadOnce/loadUniform to DiffTensor and backward pass benchmark 

### LINKED ISSUE #570: Match slangtorch backward pass performance in slangpy
The below simple benchmark compares performance between slangpy and slangtorch for a simple slang kernel. Here are the performance numbers:

`Slangpy:`
```
Forward pass: 1.497s total, 1.496726ms avg
Backward pass: 8.109s total, 8.108869ms avg
```
`Slangtorch:`
```
Forward pass: 1.388s total, 1.388282ms avg
Backward pass: 3.855s total, 3.854854ms avg
```

The goal of this bug is to ensure that slangpy and slangtorch perf numbers are alike. Attached benchmark to reproduce issue. Replace this file in `slangpy/slangpy/benchmarks` and run the benchmark as: `python test_benchmark_tensor.py`
```
import pytest
import numpy as np
from time import time
import os

import slangpy as spy
from slangpy.testing import helpers
from slangpy.testing.benchmark import BenchmarkSlangFunction

# Try to import slangtorch
try:
    import slangtorch
    HAS_SLANGTORCH = True
except ImportError:
    HAS_SLANGTORCH = False
    print("Warning: slangtorch not available, skipping slangtorch tests")

ADD_FLOATS_SLANGPY = """
import slangpy;

[Differentiable]
float add_floats(float a, float b) {
    return a + b;
}
"""

ADD_FLOATS_SLANGTORCH = """
[AutoPyBindCUDA]
[Differentiable]
[CUDAKernel]
void add_floats(DiffTensorView<float> a, DiffTensorView<float> b, DiffTensorView<float> result)
{
    uint3 dispatchIdx = cudaThreadIdx() + cudaBlockIdx() * cudaBlockDim();

    if (dispatchIdx.x >= result.size(0))
        return;

    result[dispatchIdx.x] = a[dispatchIdx.x] + b[dispatchIdx.x];
}
"""

def benchmark_torch_slangpy(
    device_type: spy.DeviceType
):
    print("\n=== Slangpy Benchmark ===")

    import torch

    # Add slangpy module search path
    slangpy_path = os.path.join(os.path.dirname(spy.__file__), "slang")

    # Create compiler options with include path
    compiler_options = spy.SlangCompilerOptions()
    compiler_options.include_paths = [slangpy_path]

    # Create device with compiler options
    device = spy.Device(type=device_type, compiler_options=compiler_options)

    # Write test code to temporary file
    test_file = "/tmp/slangpy_add_floats.slang"
    with open(test_file, "w") as f:
        f.write(ADD_FLOATS_SLANGPY)

    module = spy.Module.load_from_file(device, test_file)
    func = module.add_floats

    BUFFER_SIZE = 64*1024*1024

    a = torch.randn((BUFFER_SIZE,), dtype=torch.float32, device="cuda", requires_grad=True)
    b = torch.randn((BUFFER_SIZE,), dtype=torch.float32, device="cuda", requires_grad=True)

    # Correctness check - forward pass
    result = func(a, b)
    device.wait_for_idle()
    expected = a + b
    torch.testing.assert_close(result, expected, rtol=1e-5, atol=1e-8)
    print("✓ Forward pass correctness verified")

    # Correctness check - backward pass
    a.grad = None
    b.grad = None
    result = func(a, b)
    result.sum().backward()
    device.wait_for_idle()
    # Gradient of sum(a + b) w.r.t. a should be all ones
    torch.testing.assert_close(a.grad, torch.ones_like(a), rtol=1e-5, atol=1e-8)
    torch.testing.assert_close(b.grad, torch.ones_like(b), rtol=1e-5, atol=1e-8)
    print("✓ Backward pass correctness verified")

    # Warmup + wait for cuda
    result = func(a, b)
    device.wait_for_idle()

    # Forward pass benchmark
    test_start = time()
    for _ in range(1000):
        result = func(a, b)
    submit_end = time()
    device.wait_for_idle()
    test_end = time()

    print(f"Forward pass: {test_end - test_start:.3f}s total, {1000*(test_end - test_start)/1000:.6f}ms avg")

    # Backward pass benchmark
    # Compute forward pass once and retain graph for multiple backward passes
    result = func(a, b)
    device.wait_for_idle()
    test_start = time()
    for _ in range(1000):
        result.sum().backward(retain_graph=True)
        a.grad.zero_()
        b.grad.zero_()
    device.wait_for_idle()
    test_end = time()

    print(f"Backward pass: {test_end - test_start:.3f}s total, {1000*(test_end - test_start)/1000:.6f}ms avg")


def benchmark_torch_slangtorch():
    if not HAS_SLANGTORCH:
        print("\n=== Slangtorch Benchmark ===")
        print("Skipped: slangtorch not available")
        return

    print("\n=== Slangtorch Benchmark ===")

    import torch
    import pathlib

    # Write slangtorch code with [AutoPyBindCUDA] attribute
    test_file = "/tmp/slangtorch_add_floats.slang"
    with open(test_file, "w") as f:
        f.write(ADD_FLOATS_SLANGTORCH)

    # Load the module (slangtorch will compile and bind automatically)
    module = slangtorch.loadModule(pathlib.Path(test_file))

    # Check if add_floats is available
    if not hasattr(module, 'add_floats'):
        print("Error: add_floats function not found in module")
        print("Available attributes:", [x for x in dir(module) if not x.startswith('_')])
        return

    BUFFER_SIZE = 64*1024*1024
    THREADS_PER_BLOCK = 256
    NUM_BLOCKS = (BUFFER_SIZE + THREADS_PER_BLOCK - 1) // THREADS_PER_BLOCK

    a = torch.randn((BUFFER_SIZE,), dtype=torch.float32, device="cuda", requires_grad=True)
    b = torch.randn((BUFFER_SIZE,), dtype=torch.float32, device="cuda", requires_grad=True)
    result = torch.zeros((BUFFER_SIZE,), dtype=torch.float32, device="cuda")

    # Correctness check - forward pass
    module.add_floats(a=a, b=b, result=result).launchRaw(
        blockSize=(THREADS_PER_BLOCK, 1, 1),
        gridSize=(NUM_BLOCKS, 1, 1)
    )
    torch.cuda.synchronize()
    expected = a + b
    torch.testing.assert_close(result, expected, rtol=1e-5, atol=1e-8)
    print("✓ Forward pass correctness verified")

    # Correctness check - backward pass
    a_grad = torch.zeros_like(a)
    b_grad = torch.zeros_like(b)
    result_grad = torch.ones_like(result)
    module.add_floats.bwd(
        a=(a, a_grad),
        b=(b, b_grad),
        result=(result, result_grad)
    ).launchRaw(
        blockSize=(THREADS_PER_BLOCK, 1, 1),
        gridSize=(NUM_BLOCKS, 1, 1)
    )
    torch.cuda.synchronize()
    # Gradient of sum(a + b) w.r.t. a should be all ones
    torch.testing.assert_close(a_grad, torch.ones_like(a), rtol=1e-5, atol=1e-8)
    torch.testing.assert_close(b_grad, torch.ones_like(b), rtol=1e-5, atol=1e-8)
    print("✓ Backward pass correctness verified")

    # Warmup
    module.add_floats(a=a, b=b, result=result).launchRaw(
        blockSize=(THREADS_PER_BLOCK, 1, 1),
        gridSize=(NUM_BLOCKS, 1, 1)
    )
    torch.cuda.synchronize()

    # Forward pass benchmark
    test_start = time()
    for _ in range(1000):
        module.add_floats(a=a, b=b, result=result).launchRaw(
            blockSize=(THREADS_PER_BLOCK, 1, 1),
            gridSize=(NUM_BLOCKS, 1, 1)
        )
    submit_end = time()
    torch.cuda.synchronize()
    test_end = time()

    print(f"Forward pass: {test_end - test_start:.3f}s total, {1000*(test_end - test_start)/1000:.6f}ms avg")

    # Backward pass benchmark
    a_grad = torch.zeros_like(a)
    b_grad = torch.zeros_like(b)
    result_grad = torch.ones_like(result)
    test_start = time()
    for _ in range(1000):
        module.add_floats.bwd(
            a=(a, a_grad),
            b=(b, b_grad),
            result=(result, result_grad)
        ).launchRaw(
            blockSize=(THREADS_PER_BLOCK, 1, 1),
            gridSize=(NUM_BLOCKS, 1, 1)
        )
        a_grad.zero_()
        b_grad.zero_()
    torch.cuda.synchronize()
    test_end = time()

    print(f"Backward pass: {test_end - test_start:.3f}s total, {1000*(test_end - test_start)/1000:.6f}ms avg")


if __name__ == "__main__":
    import torch

    if not torch.cuda.is_available():
        print("ERROR: CUDA not available!")
        exit(1)

    print("=" * 70)
    print("Tensor Benchmark: Slangpy vs Slangtorch")
    print("=" * 70)
    print(f"Device: {torch.cuda.get_device_name()}")
    print(f"PyTorch version: {torch.__version__}")
    print(f"Buffer size: 64M floats ({64*1024*1024*4/(1024*1024):.1f} MB)")

    benchmark_torch_slangpy(spy.DeviceType.cuda)
    benchmark_torch_slangtorch()

    print("\n" + "=" * 70)

```

### LINKED ISSUE #769: Add GPU performance tests
The goal of thist ask is to increase coverage for GPU performance tests.
 
- Add tests that cover all Tensor types.
- Add tests explicitly for all load* variants

### PR DESCRIPTION
**Pull request by mkeshavaNV**
Fixes #570
Fixes #769

Add loadOnce (plain store backward, no atomics) and loadUniform (wave-level gradient reduction) to DiffTensor/RWDiffTensor, matching the existing DiffTensorView variants.

Includes a benchmark demonstrating atomic contention in backward pass: All tests verify gradient correctness before benchmarking.

```
 Backward — Uniform access (y[tid] = w):

  ┌──────┬─────────────────┬────────────────────────┬─────────────────────┬────────────────────────────┐
  │ Size │ DiffTensor load │ DiffTensor loadUniform │ DiffTensorView load │ DiffTensorView loadUniform │
  ├──────┼─────────────────┼────────────────────────┼─────────────────────┼────────────────────────────┤
  │ 64K  │ 0.137           │ 0.010                  │ 0.137               │ 0.010                      │
  ├──────┼─────────────────┼────────────────────────┼─────────────────────┼────────────────────────────┤
  │ 1M   │ 2.144           │ 0.072                  │ 2.201               │ 0.112                      │
  ├──────┼─────────────────┼────────────────────────┼─────────────────────┼────────────────────────────┤
  │ 4M   │ 8.564           │ 0.272                  │ 8.799               │ 0.441                      │
  └──────┴─────────────────┴────────────────────────┴─────────────────────┴────────────────────────────┘

  Backward — Per-element (f(x) = x*x):

  ┌──────┬─────────────────┬─────────────────────┬─────────────────────┬─────────────────────────┐
  │ Size │ DiffTensor load │ DiffTensor loadOnce │ DiffTensorView load │ DiffTensorView loadOnce │
  ├──────┼─────────────────┼─────────────────────┼─────────────────────┼─────────────────────────┤
  │ 64K  │ 0.007           │ 0.007               │ 0.007               │ 0.007                   │
  ├──────┼─────────────────┼─────────────────────┼─────────────────────┼─────────────────────────┤
  │ 1M   │ 0.042           │ 0.038               │ 0.049               │ 0.038                   │
  ├──────┼─────────────────┼─────────────────────┼─────────────────────┼─────────────────────────┤
  │ 4M   │ 0.154           │ 0.138               │ 0.181               │ 0.135                   │
  └──────┴─────────────────┴─────────────────────┴─────────────────────┴─────────────────────────┘

```

## Summary by CodeRabbit

* **New Features**
  * Added non-atomic "load once" and wave-reduced "load uniform" access modes for differentiable tensors to optimize gradient updates.
  * Added CUDA kernels exercising broadcast and per-element backward access patterns.

* **Tests**
  * Added CUDA-only benchmarks measuring backward-pass performance across tensor sizes and device types.
  * Added unit tests validating correctness and equivalence of the new load variants and their backward gradients.

### CODE REVIEW

File: slangpy/benchmarks/test_benchmark_bwd_diff.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    n: int,
+    benchmark_slang_function: BenchmarkSlangFunction,
+) -> None:
+    """Uniform backward using DiffTensorView load() — atomic contention."""
+    device = helpers.get_torch_device(device_type)
+    module = _load_module(device)
+    func = module.require_function("broadcast_dtv")
+    _check_uniform_correctness(func, device)
+    kwargs = _make_dtv_uniform_kwargs(n)
+    benchmark_slang_function(device, func.bwds, **kwargs)
```
[the remark this answers is not in the document]
Comment (mkeshavaNV) [reply slangpy/benchmarks/test_benchmark_bwd_diff.py:241]: Seems like this is not needed.

File: slangpy/benchmarks/test_benchmark_bwd_diff.slang [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+[CUDAKernel]
+[Differentiable]
+void square_dtv_once(uint count, DiffTensorView<float> input, DiffTensorView<float> output)
+{
+    uint tid = cudaBlockIdx().x * cudaBlockDim().x + cudaThreadIdx().x;
+    if (tid >= count)
+        return;
+    float x = input.loadOnce(tid);
+    output.storeOnce(tid, x * x);
+}
```
[the remark this answers is not in the document]
Comment (mkeshavaNV) [reply slangpy/benchmarks/test_benchmark_bwd_diff.slang:115]: it is defined in slang
